controllers/update.go

```diff
@@ -0,0 +1,309 @@
+/*
+Copyright 2021 Red Hat, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"fmt"
+
+	devfileAPIV1 "github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha2"
+	"github.com/devfile/api/v2/pkg/attributes"
+	data "github.com/devfile/library/pkg/devfile/parser/data"
+	"github.com/devfile/library/pkg/devfile/parser/data/v2/common"
+	appstudiov1alpha1 "github.com/redhat-appstudio/application-service/api/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+)
+
+func (r *ComponentReconciler) updateComponentDevfileModel(hasCompDevfileData data.DevfileData, hasComponent appstudiov1alpha1.Component) (bool, error) {
```

Comment:
This function generally looks fine to me, but I'm wondering about its long-term maintainability? Especially as we add more fields to the Component CR.

What if we generate the updated component devfile model, and compare it against the existing one (maybe via `reflect.DeepEqual` or something similar)? And if they're not equal, we update the devfile in the status to the new one?

Once we have webhooks in place, we should be able to block any invalid changes (so we shouldn't have to worry about those sneaking in).

Reply:
The thing is `reflect.DeepEqual` will not work on struct with pointers; since two structs can have the same data but those structs may have pointers referencing diff things; so it will fail

I have tried to use `reflect.DeepEqual` before going this way

Reply:
let me try out a few reflect funcs and see it actually works

Reply:
This func now generates from Component CR; and we just do a `reflect.DeepEqual` to see the diff